### DISCUSSION TITLE
Help with tar_resources()

### DISCUSSION THREAD

**Discussion by sdechaumet** [ropensci/targets#632] [Help] [answered]
Hi !

Thanks again for this awesome package, it is incredibly useful to us :) !

I got a running targets pipeline on our internal HPC using transient workers.
The pipeline deployment on the HPC use a template file from future.batchtools and is invoked with tar_make_future(). Everything works fine and the full pipeline + job submissions are done flawlessly.

But currently, some tar_target() in the pipeline need more resources and I tried to use the resources argument to specify those. By reading the documentation, I used the following parameters :

```
  tar_target(
    peakpicking_manip, {
    ## Cool code ##
    }, packages = c("MSnbase", "xcms", "data.table", "magrittr"),
    pattern = map(iter_manip),
    iteration = "list",
    resources = tar_resources(
        future = tar_resources_future(
            resources = list(walltime = "05:00:00", pmem = "5gb")
        )
     )
  )
```

But when the jobs are submitted for this target (and it is), the resources argument are not taken into account. It looks like the resources argument is not transfered to future.

My batchtools template is similar to the documentation on future.batchtools and works flawlessly (shown below if needed):

```
#!/bin/bash
#PBS -N <%= if (exists("job.name", mode = "character")) job.name else job.hash %>
#PBS -o <%= log.file %>
#PBS -j oe
#PBS -q lowprio
<% if (length(resources) > 0) {
  opts <- unlist(resources, use.names = TRUE)
  opts <- sprintf("%s=%s", names(opts), opts)
  opts <- paste(opts, collapse = ",") %>
#PBS -l <%= opts %>
<% } %>

singularity exec ./XenialR_base041-ssh.sif /usr/bin/Rscript -e 'batchtools::doJobCollection("<%= uri %>")'
```

Since the template is used correctly to submit the jobs by targets, I think the problem comes from the underlying functions of target not passing the resources argument to future.batchtools and future. But I may by wrong here.

Anyway, any insight would be appreciated and since it runs on an internal HPC, I can't make a reproducible exemple for you to work on but I'm available to run any test on my side.

Thanks again !

**Comment by wlandau** [comment, marked as answer]
Looks like you and @wresch (https://github.com/ropensci/targets/discussions/562#discussioncomment-1304379) have the same issue. I can reproduce the problem without `targets` (i.e. just `future` and `future.batchtools`) and I posted an issue at https://github.com/HenrikBengtsson/future.batchtools/issues/75. I think it is more likely to be resolved there. Using an interactive browser statement, I checked that `targets` correctly forwards resources to the `resources` argument of `future::future()`:

https://github.com/ropensci/targets/blob/d7cef99120bd6642db86b2d91ae8dedcb3346b13/R/class_future.R#L108-L125

**Reply by wlandau** [reply]
You can temporarily solve the issue by downgrading to `future` version 1.21.0. Worked on my end.

```r
remotes::install_version("future", version = "1.21.0")
```

**Reply by sdechaumet** [reply]
Thanks for the quick reply, this is exactly the same problem and I'll try your work around until future get patched.

**Reply by multimeric** [reply]
Has there been any update to this, @wlandau? It seems like this `resources` argument isn't going to work in `future.batchtools` now or in the future. Can you explain what it means to specify resources as part of the plan? Is there any way to get this working with a current version of `future`, or do we still have to use this old version?

**Reply by multimeric** [reply]
And, on a related note, is the `clustermq` backend now the preferred HPC backend for targets, because (I assume) it doesn't have this issue?

**Reply by stuvet** [reply]
@multimeric I use `tar_make_future` exclusively as I *really* wanted ad-hoc provisioning on my Slurm cluster. I have no problem passing the resources argument, though maybe this is a non-Slurm thing? I'm very happy with the setup, and seems very stable across potentially thousands of long-running tasks. I'd prefer to have true array jobs as it'd vastly speed up submission for small jobs, but sadly that doesn't seem likely according to [this](https://github.com/HenrikBengtsson/future.batchtools/issues/23#issuecomment-432353357)

My version of `future` is currently at 1.26, with `future.batchtools@22e4e5d` , `batchtools@63a6f81` & `targets@ef9341f`. I believe I used the git versions of the batchtools packages to get around some Slurm-specific problems with the default `batchtools_slurm` plan setup. I've been using a version of setup for >1yr with many different versions of targets, so I don't believe the `targets` version actually matters here.


Example of my successful resources call (beware a recent update of `targets` allows partial resources objects to be specified. Yours may not).

```
library(targets)
library(future)
library(future.batchtools)

slurm_template <- 'my.slurm.tmpl'

slurm_mcmc <- future::tweak(
  batchtools_slurm,
  template = slurm_template,
  resources = list(
    workers  = 20L,
    walltime = 3600 * 8,
    memory = 29600,
    ncpus = 16,
    ntasks = 1,
    nodes = 1,
    chunks.as.arrayjobs = TRUE, # Specific Slurm workaround - see https://github.com/ropensci/targets/discussions/570#discussion-3475914
    partition = 'ph32'
  )
)


slurm  <- future::tweak(
  batchtools_slurm,
  template = slurm_template,
  resources = list(
    template = slurm_template,
    workers = 30L,
    walltime = 720,
    memory = 7200,
    ncpus = 2,
    ntasks = 1,
    nodes = 1,
    chunks.as.arrayjobs = TRUE, # Specific Slurm workaround - see https://github.com/ropensci/targets/discussions/570#discussion-3475914
    partition = 'ph4'
  )
)

future::plan(slurm)

tar_option_set(
  resources = tar_resources(
    gcp = tar_resources_gcp(bucket = "my_gcp_bucket", predefined_acl = 'bucketLevel'),
    future = tar_resources_future(
     plan = future::plan(slurm) # Set the 2-core machine as default
    )
  ),
  tidy_eval = TRUE,
  memory='transient',
  garbage_collection = TRUE,
  deployment = 'main',
  retrieval = 'main',
  storage='main',
  format = 'qs',
  repository = 'gcp'
)

list(
 tar_target(
       my_target_1,
       ...,
       deployment = 'worker', # Will use default 2-core machine from tar_option_set
       storage = 'worker',
       retrieval = 'worker'
  ),
 tar_target(
       my_target_2,
       ...,
       deployment = 'worker',
       storage = 'worker',
       retrieval = 'worker',
       resources = tar_resources(
        future = tar_resources_future(
          plan = slurm_mcmc  # Will use the 16-core (non-threaded) machine.
        )
      )
  )
)

```

My slurm template file, in case it helps:

```
#!/bin/bash

## Job Resource Interface Definition
##
## ntasks [integer(1)]:       Number of required tasks,
## ncpus [integer(1)]:        Number of required cpus per task
## walltime [integer(1)]:     Walltime for this job, in seconds.
##                            Must be at least 60 seconds for Slurm to work properly.
## memory   [integer(1)]:     Total memory in megabytes for the job.
## mem_per_cpu   [integer(1)]:     Memory per cpu in megabytes. Takes priority over memory if both are specified.

<%
log.file = fs::path_expand(log.file)
%>

<%
number.tasks = ifelse(is.null(resources$ntasks), 1, resources$ntasks)
%>

<%
number.nodes = ifelse(is.null(resources$nodes), 1, resources$nodes)
%>
#SBATCH --job-name=<%= job.name %>
#SBATCH --output=<%= log.file %>
#SBATCH --error=<%= log.file %>
#SBATCH --time=<%= ceiling(resources$walltime / 60) %>
#SBATCH --ntasks=<%= number.tasks %>
#SBATCH --cpus-per-task=<%= resources$ncpus %>
<%= if (!is.null(resources$mem_per_cpu)) sprintf("#SBATCH --mem-per-cpu=%i", resources$mem_per_cpu) else sprintf("#SBATCH --mem=%i", resources$memory) %>
<%= if (!is.null(resources$nodes)) sprintf("#SBATCH --nodes=%i", resources$nodes) else "" %>
<%= if (!is.null(resources$partition)) sprintf(paste0("#SBATCH --partition='", resources$partition, "'")) %>
<%= if (array.jobs) sprintf("#SBATCH --array=1-%i", nrow(jobs)) else "" %>

## Initialize work environment like
## source /etc/profile
## module add ...

## Export value of DEBUGME environemnt var to slave
export DEBUGME=<%= Sys.getenv("DEBUGME") %>

<%= sprintf("export OMP_NUM_THREADS=%i", resources$omp.threads) -%>
<%= sprintf("export OPENBLAS_NUM_THREADS=%i", resources$blas.threads) -%>
<%= sprintf("export MKL_NUM_THREADS=%i", resources$blas.threads) -%>

## Run R:
## we merge R output with stdout from SLURM, which gets then logged via --output option
Rscript -e 'batchtools::doJobCollection("<%= uri %>")'


```

**Reply by multimeric** [reply]
Thanks for the info, but can you double check that it's actually respecting the `tar_resources` and isn't just using the `batchtools_slurm` defaults? Because this thread and my experience implies that it doesn't.

**Reply by stuvet** [reply]
A fair question, but it definitely does. My larger targets simply couldn't run with the defaults (they need far more cpu, memory & walltime, and I have a range of other `future::tweak` objects set up for appropriate uses, not just the two shown), & I regularly monitor `squeue` to see the different tasks appearing on the appropriate partitions. When I was trying to generate multithreading within `Stan` programs I was also regularly logging onto the workers to check CPU usage for each task was reasonably saturating the workers (1 `slurm_mcmc` task = 4 threads for each of 4 parallel chains, with one process per chain).

I remember being frustrated trying to understand the intricacies of passing the resource definitions to `tar_make_future`. This way works, as does specifying the full resource object for each target, but passing a partial (with older versions of `targets` that required a full resource object or nothing) or malformed resource object will always trigger defaults. From memory I believe the same is true if you pass the above `slurm_mcmc` objects as `plan = future::plan(slurm_mcmc)` to override the default behaviour for individual targets rather than `plan = slurm_mcmc` as shown.

Even with the later versions of `targets` I believe you'd still have to supply `future::tweak` objects with complete resources definitions to the `plan` argument of `tar_resources_future`, just that you could skip some of the arguments to the `tar_resources` object - e.g. you could leave out `gcp` or `aws`.

Currently I don't believe my `workers` arguments do anything at the per-task level, and I remember seeing that expressed as a feature request at some point, though I understand it's either v tough or impossible with the current stack.

**Reply by stuvet** [reply]
Oh & ps `batchtools_slurm` isn't an object - it's a function from `future.batchtools`. Apart from the default resources set in `tar_option_set` the only other source of defaults here would be slurm itself.

**Reply by wlandau** [reply]
Thanks for such thorough comments, @stuvet.

@multimeric, as far as updates/changes since last time, it is best to supply the whole `future` plan to the `targets` resources like @stuvet does (and e.g. the `slow_fit` target in https://books.ropensci.org/targets/hpc#future-configuration).

**Reply by multimeric** [reply]
Oh okay, so I was using:
```R
    resources = tar_resources(
      future = tar_resources_future(
        resources = list(memory = 1024 * 21)
      )
    )
```

When it looks like I should be using:
```R
    resources = tar_resources(
      future = tar_resources_future(
          plan = tweak(
            resources = list(memory = 1024 * 21))
        )
      )
    )
```

So actually that's my mistake. I saw this method of providing resources but thought it was only for *old* versions of `futures` but actually it's for newer ones.
